core/vm/exit_code/exit_code.hpp

```diff
@@ -3,17 +3,70 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef CPP_FILECOIN_CORE_VM_EXITCODE_EXITCODE_HPP
-#define CPP_FILECOIN_CORE_VM_EXITCODE_EXITCODE_HPP
+#pragma once
 
 #include <boost/optional.hpp>
-
 #include "common/outcome.hpp"
 
-#define VM_ASSERT(condition)                                  \
-  if (!(condition)) {                                         \
-    return outcome::failure(ABORT_CAST(VMExitCode::kAssert)); \
-  }
+/**
+ * Aborts execution if res has error and aborts with default_error if res has
+ * error and the error is not fatal or abort
+ */
+#define REQUIRE_NO_ERROR(expr, err_code) \
+  OUTCOME_TRY(requireNoError((expr), (err_code)))
+
+/**
+ * In addition to REQUIRE_NO_ERROR allows to create res variable
+ */
+#define REQUIRE_NO_ERROR_A(res, expr, err_code) \
+  OUTCOME_TRY((res), requireNoErrorAssign((expr), (err_code)))
+
+/**
+ * Aborts execution if res is not success
+ */
+#define REQUIRE_SUCCESS(expr) OUTCOME_TRY(requireSuccess((expr)))
+
+/**
+ * In addition to REQUIRE_SUCCESS allows to create res variable
+ */
+#define REQUIRE_SUCCESS_A(res, expr) \
+  OUTCOME_TRY((res), requireSuccessAssign((expr)))
+
+/**
+ * Returns another error if res has error and the error is not fatal or
+ * abort
+ */
+#define CHANGE_ERROR(expr, err_code) \
+  OUTCOME_TRY(changeError((expr), (err_code)))
+
+/**
+ * In addition to CHANGE_ERROR allows to create res variable
+ */
+#define CHANGE_ERROR_A(res, expr, err_code) \
+  OUTCOME_TRY((res), changeErrorAssign((expr), (err_code)))
+
+/**
+ * Aborts execution with error if res has error and the error is not fatal or
+ * abort
+ */
+#define CHANGE_ERROR_ABORT(expr, err_code) \
+  OUTCOME_TRY(changeErrorAbort((expr), (err_code)))
+
+/**
+ * In addition to CHANGE_ERROR_ABORT allows to create res variable
+ */
+#define CHANGE_ERROR_ABORT_A(res, expr, err_code) \
+  OUTCOME_TRY((res), changeErrorAbortAssign((expr), (err_code)))
+
+/**
+ * Return VMExitCode as VMAbortExitCode for special handling
+ */
+#define ABORT_CAST(err_code) static_cast<VMAbortExitCode>(err_code)
```

Comment:
Add method (maybe like `asExitCode`) which will only allow casting from `VMAbortExitCode` and `VMExitCode`

Reply:
Where do we need it to use?

Reply:
Where it's used.
This macros `static_cast`'s anything to `VMAbortExitCode`, this isn't right

Reply:
Why? When spec actor aborts anything, it is aborted anyway.
If we allow to keep the current error type for non-VMExitode, then this error could be changed later.
But when we cast it to VMAbortExitCode - it will be not changed

Reply:
`outcome::result` uses `std::system_error`, which combines enum value and enum id (category), so it can distinguish same enum values from different enums.
Vm should only handle `VM(Abort)ExitCode` and `VMFatal` (can't `static_cast` to `VM(Abort)ExitCode`) and avoid `static_cast(error.value())` for all other errors (e.g. `AmtError`, `HamtError`, ...), otherwise they will map to some meaningless `VMAbortExitCode` value

Reply:
Done